lib/shelley/src/Cardano/Wallet/Shelley/Network.hs

```diff
@@ -0,0 +1,508 @@
+
+import Cardano.Chain.Slotting
+    ( EpochSlots (..) )
+import Codec.SerialiseTerm
+    ( CodecCBORTerm )
+import Network.Mux.Interface
+    ( AppType (..) )
+import Network.TypedProtocol.Channel
+    ( Channel )
+import Network.TypedProtocol.Codec
+    ( Codec )
+import Network.TypedProtocol.Codec.Cbor
+    ( DeserialiseFailure )
+import Network.TypedProtocol.Driver
+    ( TraceSendRecv, runPeer )
+import Ouroboros.Consensus.Ledger.Byron
+    ( ByronBlockOrEBB (..)
+    , GenTx
+    , decodeByronBlock
+    , decodeByronGenTx
+    , decodeByronHeaderHash
+    , encodeByronBlock
+    , encodeByronGenTx
+    , encodeByronHeaderHash
+    )
+import Ouroboros.Consensus.Ledger.Byron.Config
+    ( ByronConfig )
+import Ouroboros.Consensus.NodeId
+    ( CoreNodeId (..) )
+import Ouroboros.Network.Block
+    ( decodePoint, encodePoint )
+import Ouroboros.Network.Mux
+    ( OuroborosApplication (..) )
+import Ouroboros.Network.NodeToClient
+    ( NodeToClientProtocols (..)
+    , NodeToClientVersion (..)
+    , NodeToClientVersionData (..)
+    , connectTo
+    , nodeToClientCodecCBORTerm
+    )
+import Ouroboros.Network.Protocol.ChainSync.Client
+    ( ChainSyncClient (..)
+    , ClientStIdle (..)
+    , ClientStIntersect (..)
+    , ClientStNext (..)
+    , chainSyncClientPeer
+    )
+import Ouroboros.Network.Protocol.ChainSync.Codec
+    ( codecChainSync )
+import Ouroboros.Network.Protocol.ChainSync.Type
+    ( ChainSync )
+import Ouroboros.Network.Protocol.Handshake.Version
+    ( DictVersion (..), simpleSingletonVersions )
+import Ouroboros.Network.Protocol.LocalTxSubmission.Client
+    ( LocalTxSubmissionClient (..), localTxSubmissionClientPeer )
+import Ouroboros.Network.Protocol.LocalTxSubmission.Codec
+    ( codecLocalTxSubmission )
+import Ouroboros.Network.Protocol.LocalTxSubmission.Type
+    ( LocalTxSubmission )
+
+import Cardano.Crypto
+    ( ProtocolMagicId (..) )
+import Control.Monad.Class.MonadST
+    ( MonadST )
+import Control.Monad.Class.MonadSTM
+    ( MonadSTM
+    , TQueue (..)
+    , atomically
+    , newEmptyTMVarM
+    , newTQueue
+    , putTMVar
+    , readTQueue
+    , takeTMVar
+    , writeTQueue
+    )
+import Control.Monad.Class.MonadThrow
+    ( MonadThrow )
+import Control.Monad.Class.MonadTimer
+    ( MonadTimer, threadDelay )
+import Control.Monad.Trans.Except
+    ( ExceptT (..) )
+import Control.Tracer
+    ( Tracer, contramap )
+import Data.ByteString.Lazy
+    ( ByteString )
+import Data.Text
+    ( Text )
+import Data.Void
+    ( Void )
+import Network.Socket
+    ( AddrInfo (..), Family (..), SockAddr (..), SocketType (..) )
+
+import qualified Codec.Serialise as CBOR
+import qualified Network.Socket as Socket
+import qualified Ouroboros.Network.Block as Abstract
+
+--------------------------------------------------------------------------------
+--
+-- Stateless Interface
+--
+-- This is a first attempt at using the ChainSync mini protocol in a stateless
+-- manner. Note that this is a bit unfortunate as the
```

Comment:
> Note that this is a bit unfortunate as the

What comes next? :)

Reply:
:no_mouth: ... At this point, I am afraid to have forgotten. 

I probably wanted to outline the fact that, the mini-protocols are intrinsically _stateful_ but our network layer is currently stateless. So here, I've turned the mini-protocols into a stateless API, although, we might want to do it the other way around and make our network stateful :) 

Instead of passing around the block header hash to get the next blocks, we could let the network layer keep track of where the pointer is. We'd need a way to resolve an initial intersection of course when spinning up a network layer. 

Having said that, I think it's quite an important design change that will really pay off when dealing with rollback. So, in this first iteration, we may simply use the mini-protocols in a stateless fashion and review our network layer design to work statefully (including for Jormungandr) when we start working on rollbacks.